src/main/java/com/google/opensesame/projects/ProjectQuery.java

```diff
@@ -0,0 +1,118 @@
+   * @return Returns the list of QueryFilters or null if an error was encountered while parsing.
+   * @throws ServletValidationException
+   */
+  private static List<QueryFilter> getQueryFiltersFromRequest(HttpServletRequest request)
+      throws IOException, ServletValidationException {
+    String[] filterRequests = request.getParameterValues(FILTER_QUERY_PARAM);
+    if (filterRequests == null) {
+      return Arrays.asList();
+    }
+
+    List<QueryFilter> queryFilters = new ArrayList<QueryFilter>();
+    for (String filterRequest : filterRequests) {
+      QueryFilter queryFilter = QueryFilter.fromString(filterRequest);
+      queryFilters.add(queryFilter);
+    }
+
+    return queryFilters;
+  }
+
+  /**
+   * Queries for ProjectEntities in the datastore based on a servlet request.
+   *
+   * @param request The servlet request defining the query.
+   * @param response The servlet response to send errors to.
+   * @return Returns a collection of ProjectEntities from the query or null if the query was
+   *     invalid.
+   * @throws IOException
+   */
+  public static Collection<ProjectEntity> queryFromRequest(HttpServletRequest request)
+      throws IOException, ServletValidationException {
+    String[] projectIds = request.getParameterValues(ProjectEntity.PROJECT_ID_PARAM);
+    Collection<ProjectEntity> projectEntities;
+    if (projectIds != null) {
+      projectEntities = ofy().load().type(ProjectEntity.class).ids(projectIds).values();
+
+      if (projectIds.length != projectEntities.size()) {
+        throw new ServletValidationException(
+            "One or more of the supplied IDs is invalid and those projects could not be found.",
+            "One or more of the requested projects does not exist.",
+            HttpServletResponse.SC_NOT_FOUND);
+      }
+    } else {
+      // TODO(Richie): Add pagination support.
+      // TODO(Richie): Add ordering support.
+      List<QueryFilter> queryFilters = getQueryFiltersFromRequest(request);
+      if (queryFilters.size() > 0) {
+        List<ProjectEntity> intermediateIntersection = null;
+
+        for (QueryFilter queryFilter : queryFilters) {
+          List<ProjectEntity> filteredProjects =
+              ofy()
+                  .load()
+                  .type(ProjectEntity.class)
+                  .filter(queryFilter.condition, queryFilter.comparisonObject)
+                  .list();
+
+          if (intermediateIntersection != null) {
+            intermediateIntersection =
+                filteredProjects.stream()
+                    .filter(intermediateIntersection::contains)
+                    .collect(Collectors.toList());
+          } else {
+            intermediateIntersection = filteredProjects;
+          }
+        }
+
+        projectEntities = intermediateIntersection;
+      } else {
+        projectEntities = ofy().load().type(ProjectEntity.class).list();
+      }
+    }
+
+    return projectEntities;
```

Comment:
It seems like there is a lot going on in this function, so I think it should be broken up into smaller functions:
- a function that gets `ProjectEntity`s from projectIds (in my example I call it `getProjectEntitiesByIds`)
- a function that gets `ProjectEntity`s from a queryFilters (in my example I call it `getProjectEntitiesByQueryFilters`)
```suggestion
    String[] projectIds = request.getParameterValues(ProjectEntity.PROJECT_ID_PARAM);
    if (projectIds != null) {
      return getProjectEntitiesByIds(projectIds);
    } else {
      List<QueryFilter> queryFilters = getQueryFiltersFromRequest(request);
      return getProjectEntitiesByQueryFilters(queryFilters);
    }
```

Reply:
I agree with you. I just implemented this change and will push shortly. Thanks for all of the help!